src/psyclone/gocean1p0.py

```diff
@@ -460,6 +465,101 @@ def __init__(self, parent=None,
         if not GOLoop._bounds_lookup:
             GOLoop.setup_bounds()
 
+    # -------------------------------------------------------------------------
+    def _halo_read_access(self, arg):
+        '''Determines whether the supplied argument has (or might have) its
+        halo data read within this loop. Returns True if it does, or if
+        it might and False if it definitely does not.
+
+        :param arg: an argument contained within this loop
+        :type arg: :py:class:`psyclone.dynamo0p3.DynArgument`
+        :return: True if the argument reads, or might read from the \
+                 halo and False otherwise.
+        :rtype: bool
+
+        '''
+        if arg.type != 'field':
+            return False
+        else:
+            return True
+
+    def unique_fields_with_halo_reads(self):
+        ''' Returns all fields in this loop that require at least some
+        of their halo to be clean to work correctly. '''
+
+        unique_fields = []
+        unique_field_names = []
+
+        for call in self.kernels():
+            for arg in call.arguments.args:
+                if self._halo_read_access(arg):
+                    if arg.name not in unique_field_names:
+                        unique_field_names.append(arg.name)
+                        unique_fields.append(arg)
+        return unique_fields
+
+    def create_halo_exchanges(self):
+        '''Add halo exchanges before this loop as required by fields within
+        this loop. To keep the logic simple we assume that any field
+        that accesses the halo will require a halo exchange and then
+        remove the halo exchange if this is not the case (when
+        previous writers perform sufficient redundant computation). It
+        is implemented this way as the halo exchange class determines
+        whether it is required or not so a halo exchange needs to
+        exist in order to find out. The appropriate logic is coded in
+        the _add_halo_exchange helper method. '''
+        for halo_field in self.unique_fields_with_halo_reads():
+            # for each unique field in this loop that has its halo
+            # read (including annexed dofs), find the previous update
+            # of this field
+            prev_arg_list = halo_field.backward_write_dependencies()
+            if not prev_arg_list:
+                # field has no previous dependence so create new halo
+                # exchange(s) as we don't know the state of the fields
+                # halo on entry to the invoke
+                self._add_halo_exchange(halo_field)
+            else:
+                # field has one or more previous dependencies
+                if len(prev_arg_list) > 1:
+                    # field has more than one previous dependencies so
+                    # should be a vector
+                    if halo_field.vector_size <= 1:
```

Comment:
I (think) this is LFRic specific - we don't have vectors in GOcean.

Reply:
I deleted the `vector_size` references from GOcean. This did allowed me to simplify the GOcean HaloEx code, but it prevented me to generalise some of these methods as I initially expected.